src/Pretty.hs

```diff
@@ -241,6 +241,16 @@ instance Pretty VarInfo core => Pretty VarInfo (ScopedList core) where
 instance PrettyBinder VarInfo CompleteDecl where
   ppBind env (CompleteDecl d) = ppBind env d
 
+instance PrettyBinder VarInfo [CompleteDecl] where
+  ppBind env [] = (text "no-decls", env)
+  ppBind env [decl] = ppBind env decl
+  ppBind env decls =
+    let (docs, envs) = unzip $ fmap (ppBind env) decls
+        doc = align $ group $ vsep docs
+        env' = mconcat envs
+    in (text "group" <> doc, env')
```

Comment:
At first I incorrectly assumed that the `env` was supposed to be threaded like a `State`, but I later realized that the input `Env` is like a `Reader` and the output `Env` is like a `Writer`. I hope I got it right this time; what would be a good way to test this implementation?

Reply:
I don't suspect that this is correct, because each decl should be in scope for the ones just after it. In other words, this would make it a foldr-like computation, rather than an fmap-like.

A way to test is to have a list with two definitions, where the second refers to the first, and then see if the var in the pretty-printed second one shows up as an internal thing with !s, or as the actual name.

At a higher level, why would the core language need pretty-printing like this? As of now, it's not attempting to mimic the surface language, but instead looks a bit more like ML, just to keep them conceptually separate.

If this is called for the body of a completed module and the body of a Meta, then I would think that a quick foldr surrounded by a vsep would be sufficient, rather than this instance.

Reply:
Indeed, this implementation is quite wrong. With a few cosmetic fixes, it pretty-prints `group.kl` like this:

```
module "\"/Users/gelisam/working/haskell/klister/examples/group.kl\""
    import #[group.kl:2.16-2.22]<"kernel"> ⇑ 1
    define a : Signal := Signal 1
    meta
        define a : Signal := Signal 10
    define b : Signal := a
    define c : Signal := b
    meta
        define b : Signal := a
        define c : Signal := !!(Var 117)!!
    example@group.kl:14.1-15.1 a : Signal
    example@group.kl:15.1-16.1 b : Signal
    example@group.kl:16.1-17.1 c : Signal
    meta
        example@group.kl:18.3-19.3 a : Signal
        example@group.kl:19.3-20.3 b : Signal
        example@group.kl:20.3-20.14 c : Signal
```

That is, it contains the dreaded exclamation marks. Replacing the implementation with a `foldr` fixes those exclamation marks... but introduces more elsewhere! Any idea what's going on?

```
module "\"/Users/gelisam/working/haskell/klister/examples/group.kl\""
    import #[group.kl:2.16-2.22]<"kernel"> ⇑ 1
    define a : Signal := Signal 1
    meta
        define a : Signal := Signal 10
    define b : Signal := a
    define c : Signal := b
    meta
        define b : Signal := a
        define c : Signal := b
    example@group.kl:14.1-15.1 a : Signal
    example@group.kl:15.1-16.1 b : Signal
    example@group.kl:16.1-17.1 c : Signal
    meta
        example@group.kl:18.3-19.3 a : Signal
        example@group.kl:19.3-20.3 !!(Var 117)!! : Signal
        example@group.kl:20.3-20.14 !!(Var 131)!! : Signal
```

Reply:
Ah, I finally understand what's going on! Due to its `Env -> (..., Env)` signature, I keep incorrectly assuming that it behaves like a State, by returning a version of the input Env extended with some new variables. I now realize that the output Env only includes the newly-bound variables. For this decl list, this makes things a bit awkward because we do need to thread an Env through, but that Env must only contain the newly-bound variables, so it must start with `mempty`, not with the input Env. That input env must then be added to the accumulated Env each time we call `ppEnv` on one of the decls. After that we can accumulate the variables newly-bound by this decl into the Env we are threading through.